src/python.cpp

```diff
@@ -1251,110 +1251,140 @@ SEXP py_to_r(PyObject* x, bool convert) {
-    if (param_name.is_null())
-      throw PythonException(py_fetch_error());
+  while ((item = PyIter_Next(parameters_iterator))) // new ref
+  {
+    PyObjectPtr item_(item);
+    PyObject *name = PyTuple_GetItem(item, 0);  // borrowed reference
+    PyObject *param = PyTuple_GetItem(item, 1); // borrowed reference
 
-    PyObjectPtr param_kind(PyObject_GetAttrString(param.get(), "kind"));
-    if (param_kind.is_null())
-      throw PythonException(py_fetch_error());
+    PyObjectPtr kind_(PyObject_GetAttrString(param, "kind")); // new ref
+    if (kind_.is_null()) throw PythonException(py_fetch_error());
+    PyObject *kind = kind_.get();
+
+    if (kind == inspect_Parameter_VAR_KEYWORD ||
+        kind == inspect_Parameter_VAR_POSITIONAL)
+    {
+      if (!has_dots)
+      {
+          GrowList(r_args, Rf_install("..."), R_MissingArg);
+          has_dots = true;
+      }
+      continue;
+    }
 
-    PyObjectPtr param_default(PyObject_GetAttrString(param.get(), "default"));
+    if (!has_dots && kind == inspect_Parameter_KEYWORD_ONLY)
+    {
+      GrowList(r_args, Rf_install("..."), R_MissingArg);
+      has_dots = true;
+    }
+
+    SEXP arg_default = R_NilValue;
+    PyObjectPtr param_default(PyObject_GetAttrString(param, "default")); // new ref
     if (param_default.is_null())
       throw PythonException(py_fetch_error());
 
-    // If we encounter our first kw_only param
-    // without having encountered `*args` or `**kw`,
-    // we insert `...` before the actual parameter.
-    if (param_kind == kw_only && !var_encountered) {
-      formals << Named("...", R_MissingArg);
-      var_encountered = true;
+    if (param_default.get() == inspect_Parameter_empty)
+    {
+      arg_default = R_MissingArg;
     }
-
-    // If we encounter the first of `*args` or `**kw`,
-    // we insert `...` instead of a parameter.
-    // foo(*args, b=1, **kw) -> foo(..., b=1)
-    if (param_kind == var_pos || param_kind == var_kw) {
-      if (!var_encountered) {
-        formals << Named("...", R_MissingArg);
-        var_encountered = true;
-      }
-    // For a parameter w/o default value, we insert `R_MissingArg`.
-    // There is inspect.Parameter(..., *, default = Parameter.empty, ...)
-    // so we check if param_kind != kw_only for this corner case.
-    } else if (param_kind != kw_only && param_default == empty_param) {
-      formals << Named(as_utf8_r_string(param_name.get()), R_MissingArg);
-    // If we arrive here we have a parameter with default value.
-    } else {
-      // Here we could convert a subset of python objects to R defaults.
-      // Plain values (numeric, character, NULL, ...) are stored as is,
-      // variables, calls, ... are stored as `symbol` or `language`.
-      formals << Named(as_utf8_r_string(param_name.get()), R_NilValue);
+    else
+    {
+      arg_default = py_to_r(param_default, true); // callable.convert());
     }
+
+    GrowList(r_args, Rf_install(PyUnicode_AsUTF8(name)), arg_default);
```

Comment:
Nit: arguments to `Rf_install()` should be in the native encoding, which may not be UTF-8. You might want to use `translateChar()` or some similar R API to confirm you're getting a string in the native encoding.

Reply:
Thanks, I didn't know that about `Rf_install()`! I've updated the snippet to 
```cpp
const char *name_char = PyUnicode_AsUTF8(name);
SEXP name_sym = Rf_installTrChar(Rf_mkCharCE(name_char, CE_UTF8));
```